### PR TITLE
Vendor BUILD file script

### PR DESCRIPTION
**Pull request by siggisim**
A script that generates and vendors BUILD files of go repositories and prints out a corresponding http_archive rule to replace the go_repository.

I've applied the changes to two of our worst offenders:
- `com_github_docker_docker` 600+s
- `org_golang_google_api` 100+s

I also collected a list of a couple dozen others that take 20+s that we can convert if this experiment goes well.

The script is run like this:
```
./tools/vendor.sh <org/repo> <commit> [custom-import-path] [custom-name]
```

Here's an example for importing a simple github repo at a given commit:
```
./tools/vendor.sh docker/docker 363e9a88a11be517d9e8c65c998ff56f774eb4dc
```

And a more complicated one that has a custom import path and name:
```
./tools/vendor_git.sh googleapis/google-api-go-client c0067489eddd6a0c8aad7c6f9ac0ebd946c0f3d8 google.golang.org/api org_golang_google_api
```

This is useful for go repos that are slow to download, and slower to generate BUILD files for (like docker/docker which takes 600s to load).

Build files are generated using gazelle - similar to how go_repository generates them.

Build files are vendored as patches, so they can easily be applied to the http_archive that's downloaded. This is the Jay Conrod's [recommended way](https://github.com/bazelbuild/bazel-gazelle/issues/533) of modifying BUILD files post-download.

I initially wrote this to generate `git_repository` rules rather than `http_archive`, but then learned that `git_repository` downloads are [not stored in the repository cache](https://github.com/bazelbuild/bazel/issues/5116) the way that `http_archive`s are.

Here's what the generated output looks like 
```
Add this to your deps.bzl:

    # Generated with ./tools/vendor_http.sh googleapis/google-api-go-client c0067489eddd6a0c8aad7c6f9ac0ebd946c0f3d8 google.golang.org/api org_golang_google_api
    http_archive(
        name = "org_golang_google_api",
        sha256 = "0d54a64322e02492ecbf40723b2392cea959069185a7736019f74b20de752fa5",
        strip_prefix = "google-api-go-client-c0067489eddd6a0c8aad7c6f9ac0ebd946c0f3d8",
        urls = ["https://github.com/googleapis/google-api-go-client/archive/c0067489eddd6a0c8aad7c6f9ac0ebd946c0f3d8.zip"],
        patches = ["@%s//buildpatches:org_golang_google_api" % workspace_name],
        patch_args = ["-s", "-p0"],
    )
```

---

**Version bump**: None

### CODE REVIEW

File: tools/vendor.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        name = \"$custom_name\",
+        sha256 = \"$sha256\",
+        strip_prefix = \"${root_dir%/}\",
+        urls = [\"$archive_url\"],
+        patches = [\"@%s//buildpatches:$custom_name\" % workspace_name],
+        patch_args = [\"-s\", \"-p0\"],
+    )
+"""
+
+# Remove our temporary directory
```
Comment (bduffany) [inline tools/vendor.sh:67]: nit: you can put this logic in a cleanup block just after creating the directory, so it is more likely to get cleaned up even if the script fails or if the user quits with Ctrl+C

```bash
cleanup() {
  rm -rf "$pwd/.buildbuddy-vendor-temp"
}
trap cleanup EXIT
```

File: tools/vendor.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+fi
+
+pwd=$(pwd)
+
+# Make sure directories exist
+mkdir -p buildpatches
+mkdir -p .buildbuddy-vendor-tmp
+cd .buildbuddy-vendor-tmp
+
+reponame=$(basename "$1" .git)
```
Comment (bduffany) [inline tools/vendor.sh:25]: TIL basename takes a second arg to strip a suffix :scream:

File: tools/vendor.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    example: ./tools/vendor_git.sh googleapis/google-api-go-client c0067489eddd6a0c8aad7c6f9ac0ebd946c0f3d8 google.golang.org/api org_golang_google_api
+    """
+    exit
+fi
+
+pwd=$(pwd)
+
+# Make sure directories exist
+mkdir -p buildpatches
+mkdir -p .buildbuddy-vendor-tmp
```
Comment (bduffany) [inline tools/vendor.sh:22]: nit, consider `tmpdir=$(mktemp -d)` to avoid putting stuff in the workspace

File: tools/vendor.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+# Go into the unzipped directory
+root_dir=$(unzip -Z -1 $reponame.zip | head -1)
+cd $root_dir
+
+# Fallback to github repo as prefix
+github_prefix="github.com/$1"
+custom_prefix=${3-$github_prefix}
+
+# Fallback to github repo as name
+github_name="com_github_${1/\//_}"
```
Comment (bduffany) [inline tools/vendor.sh:46]: I think gazelle replaces `-` with `_` as well

File: tools/vendor.sh (hunk 1 above)
Comment (siggisim) [reply tools/vendor.sh]: Great idea, done!

File: tools/vendor.sh (hunk 2 above)
Comment (siggisim) [reply tools/vendor.sh:33]: Magic 🧙

File: tools/vendor.sh (hunk 3 above)
Comment (siggisim) [reply tools/vendor.sh]: Awesome suggestion - done!

File: tools/vendor.sh (hunk 4 above)
Comment (siggisim) [reply tools/vendor.sh]: Ah yeah, I hadn't thought about this too much - switched to all non alphanumeric characters.

File: tools/vendor.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
+set -e
```
Comment (bduffany) [inline tools/vendor.sh:1]: missed this earlier - might need a shebang line to make this execute w/ bash

Comment (siggisim) [reply tools/vendor.sh:2]: Done!

**Review by tylerwilliams — APPROVED** [review]
LGTM

Dope!

**Pull request merged**